torchdata/nodes/samplers/multi_node_weighted_sampler.py

```diff
@@ -0,0 +1,216 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+import copy
+from typing import Any, Dict, Mapping, Optional
+
+import torch
+from torchdata.nodes.base_node import BaseNode, T
+from torchdata.nodes.samplers.utils import StopCriteria
+
+from .utils import _get_rank_seed, get_rank_and_world_size
+
+
+class MultiNodeWeightedSampler(BaseNode[T]):
+    """A node that samples from multiple datasets with weights."""
+
+    DATASET_NODE_STATES_KEY = "dataset_node_states"
+    NUM_YIELDED_KEY = "num_yielded"
+    WEIGHTED_SAMPLER_STATE_KEY = "weighted_sampler_state"
+    DATASETS_EXHAUSTED_KEY = "datasets_exhausted"
+
+    def __init__(
+        self,
+        source_nodes: Mapping[str, BaseNode[T]],
+        weights: Dict[str, float],
+        stop_criteria: str = StopCriteria.CYCLE_UNTIL_ALL_DATASETS_EXHAUSTED,
+        rank: int | None = None,
+        world_size: int | None = None,
+        seed: int = 0,
+    ) -> None:
+        super().__init__()
+        self.source_nodes = source_nodes
+        self.weights = weights
+        self.stop_criteria = stop_criteria
+        self.dataset_names = list(self.source_nodes.keys())
+        self._num_yielded = 0
+        self.seed = seed
+
+        if rank is None or world_size is None:
+            self.rank, self.world_size = get_rank_and_world_size()
+        else:
+            self.rank = rank
+            self.world_size = world_size
+
+        self.epoch = 0
+
+        self._validate()
+
+    def _validate(self) -> None:
+        if self.stop_criteria not in [
+            StopCriteria.CYCLE_UNTIL_ALL_DATASETS_EXHAUSTED,
+            StopCriteria.ALL_DATASETS_EXHAUSTED,
+            StopCriteria.FIRST_DATASET_EXHAUSTED,
+        ]:
+            raise ValueError(
+                f"Invalid {self.stop_criteria=}. stop_criteria must be one of: CYCLE_UNTIL_ALL_DATASETS_EXHAUSTED, FIRST_DATASET_EXHAUSTED, ALL_DATASETS_EXHAUSTED"
+            )
+
+        # Check if keys of source_nodes and weights are the same
+        if set(self.dataset_names) != set(self.weights.keys()):
+            raise ValueError(
+                f"Invalid {self.weights=}. For multi-dataset weighted sampling, keys of source_nodes and weights must be the same",
+            )
+
+        for weight in self.weights.values():
+            if not isinstance(weight, float) or weight <= 0:
+                raise ValueError(
+                    f"Invalid {self.weights=}. For multi-dataset weighted sampling, weights must be a 1d sequence, non-negative, and non-zero"
```

Comment:
It'd be good for user understanding to have either a brief comment, (or include directly in the error message) about why this might unintentionally cause infinite loops